documentation[next]: CODING_GUIDELINES.md edits (#1272)
GridTools/gt4py: CODING_GUIDELINES.md

Temporarily, tests for testing utilities can be placed next to the module containing them, with the name `test_util_<module>.py`. This should be taken as a hint that the tested utils should be moved into the library.

#### Integration Tests Utils

Integrations tests come with their own utilities, found in `cases.py`, for better test automation and simplification:

> - Field type annotations, e.g. `IJKField = Field[[IDim, JDim, KDim], np.int64]`.

Reviewer: ```suggestion
- Predefined field type annotations, e.g. `IJKField = Field[[IDim, JDim, KDim], np.int64]`.
```

- Test fixtures: `cartesian_case` for structured, e.g. `IDim` x `JDim`; `unstructured_case`, e.g. `EdgeDim`.
- Parameter allocations for objects, derived directly from the decorator's function definition:
  ```
     input_param = cases.allocate(case_fixture, decorator_name, "input_label")()
  ```
- Data initialization/modification classes, such as `ConstInitializer()`, to be used as `allocate(...).strategy()` parameters
- Decorator verification functions:
  - `cases.verify()`: used when input and output field generation cannot be automated. For example the case when the `extend()` method needs to be used for an offset or sparse fields are part of the input dataset.
  - `cases.verify_with_default_data()`: used when input and output fields generation can be automated.

> - Backends are set automatically with default switched off. However, if explication is needed, they can be explicated as cases attributes.

Author: I hope it's more clear now

Temporarily, tests for testing utilities can be placed next to the module containing them, with the name `test_util_<module>.py`. This should be taken as a hint that the tested utils should be moved into the library.

#### Integration Tests Utils

> Integrations tests come with their own utility functions, found in `cases.py`, for better test automation and simplification:

Reviewer 2: ```suggestion
Integrations tests come with their own utilities, found in `cases.py`, for better test automation and simplification:
```
'Cause the first thing listed is already not a function.

> Integrations tests comprise their own utility functions for features' testing, found in `cases.py` for better test automation and simplification:

Reviewer 2: ```suggestion
Integrations tests come with their own utility functions, found in `cases.py`, for better test automation and simplification:
```

>     - Fields definitions, e.g. IJKField = Field[[IDim, JDim, KDim], np.int64]:

Reviewer 2: ```suggestion
    - Field type annotations, e.g. `IJKField = Field[[IDim, JDim, KDim], np.int64]`.
```

>     - Cases fixtures: cartesian_case for structured, e.g. IDim, JDim; unstructured_case, e.g. EdgeDim.

Reviewer 2: ```suggestion
    - Test fixtures: `cartesian_case` for structured, e.g. `IDim` x `JDim`; `unstructured_case`, e.g. `EdgeDim`.
```

>     - Parameters allocations to objects, derived directly from the decorator's function definition:

Reviewer 2: ```suggestion
    - Parameter allocations for objects, derived directly from the decorator's function definition:
```

It's not very clear what parameter allocation actually means, at least not from the guidelines.

        ```
           input_param = cases.allocate(case_fixture, decorator_name, "input_label")()
        ```

>     - Data initializations/modifications classes, such as ConstInitializer(), to be used as `allocate(...).strategy()` parameters

Reviewer 2: ```suggestion
    - Data initialization/modification classes, such as `ConstInitializer()`, to be used as `allocate(...).strategy()` parameters
```

>     - Decorators verification functions:

Reviewer 2: ```suggestion
    - Decorator verification functions:
```

>         - `cases.verify()`: used when input and output fields generation cannot be automated. This is for example the case when the `extend()` method needs to be used for an offset or sparse fields are part of the input dataset:

Reviewer 2: ```suggestion
        - `cases.verify()`: used when input and output field generation cannot be automated. For example the case when the `extend()` method needs to be used for an offset or sparse fields are part of the input dataset.
```

        - `cases.verify_with_default_data()`: used when input and output fields generation can be automated.

>     - Backends are set automatically with default switched off. However, if explication is needed, they can be explicated as cases attributes.

Reviewer 2: Can you improve the phrasing of this? What do you mean by default being switched off? By "explication" you mean overriding?

> In case new test utils need to be written, it is important to first check whether any utilities can be use or modified.

Reviewer 2: ```suggestion
In case new test utils need to be written, it is important to first check whether any utilities can be used or modified.
```

> If this is not the case but the new utility supports similar functionalities as others, it should be placed in the same file. Naming convention should mirror other cases and reflect its purpose.

Reviewer 2: ```suggestion
If this is not the case but the new utility supports similar functionalities as others, it should be placed in the same file. Naming convention should mirror other cases and reflect the purpose.
```

> In case that the new utility is a completely stand-alone component, a new file shall be created in the mutual directory of all usages.

Reviewer 2: ```suggestion
In case that the new utility is a completely stand-alone component, a new file should be created in the bottom-most directory enclosing all usages of the utility.
```
I'm not sure if this was the intended meaning.

> In case new test utils need to be written, it is important to first check whether any utilities can be used or modified.
> If this is not the case but the new utility supports similar functionalities as others, it should be placed in the same file. Naming convention should mirror other cases and reflect the purpose.
> In case that the new utility is a completely stand-alone component, a new file should be created in the bottom-most directory enclosing all usages of the utility.

Reviewer: ```suggestion
In case new features are needed, check if an existing utility can be extended before implementing a whole new one. The name of new utilities should reflect their purpose and follow the naming conventions of existing tools. If the new utility is a completely stand-alone component, create a new file in the bottom-most directory enclosing all usages of the utility; otherwise, place it in the appropriate file containing utilities with similar functionalities.
```

> Note: The name cases for the new test module was based on the idea that details like backend, grid size etc should be summarized in a parametrizable “test case” (there being two types, cartesian and unstructured ones).

Reviewer 2: ```suggestion
Note: The name `cases` for the new test module was chosen based on the idea that details like backend, grid size, etc. should be summarized in a parametrizable “test case”, of which there are two major types: cartesian and unstructured.
```
Is the grid type a parameter in the test case? If so, why is it not listed alongside backend, grid size, etc., if not, then why is grid type mentioned as test type?

> Note: The name cases for the new test module was based on the idea that details like backend, grid size etc should be summarized in a parametrizable “test case” (there being two types, cartesian and unstructured ones).\_

Author: ```suggestion
Note: The name cases for the new test module was based on the idea that details like backend, grid size etc should be summarized in a parametrizable “test case” (there being two types, cartesian and unstructured ones).
```

> Note: The name `cases` for the new test module was chosen based on the idea that details like backend, grid size, etc. should be summarized in a parametrizable “test case”, of which there are two major types: cartesian and unstructured.

Reviewer 2: ```suggestion
Note: The name `cases` for the new test module was chosen based on the idea that details like backend, grid size, etc. should be summarized in a parametrizable “test case”, of which there are two types: cartesian and unstructured.
```
Yeah okay, my previous suggestion was not good either.

> TODO: add missing test conventions.

Author: Iterator tests still need to be updated, I am not sure if there will be more conventions to add to this file then
